### PR TITLE
Week 4 assessment heather

### PR DESCRIPTION
**Pull request by hroulston**

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -27,31 +44,148 @@ var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saf
...
+//     }else{
+//       return no_dups.push(all_nums[i])
+//     }
+//   }
+// }
+//
+// console.log(no_duplicates(testArray1, testArray2))
+
+// Well this isn't working but I did find a solution that was on my thought process but makes a little more sense so I am going to try it. I also learned a little more about the different loops which was fun. I didn't understand the syntax of the for of loop originally but after some research I learned how handy this loop as well as the while and for in loops. This is basically stating if the value found in the all_nums array is not in the no_dups array then push it into the no_dups array.
+// I also realized after finding the for loop that my need to sort the all_nums array was unncessary.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:179]: Great work on your research here!

File: code-challenges.test.js [hunk 2]
```diff
@@ -27,31 +44,148 @@ var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saf
...
+// Then I am going to take index [0] and push it to the final_array
+// Next I will do array[arr.length-1] to grab the last index and push that to the final_array
+// Return final_array
+
+const min_max = (array) => {
+  final_array = []
+  sorted_array = array.sort((a,b)=>a-b)
+  final_array.push(sorted_array[0], sorted_array[sorted_array.length-1])
+  return final_array
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:127]: well done!

File: code-challenges.test.js [hunk 3]
```diff
@@ -18,6 +18,23 @@
...
+
+    // expect(shuffle(colors1)).not.toContain("purple")
+    expect(shuffle(colors1)).toContain("blue", "green", "yellow", "pink")
+
+    var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]
+
+    // expect(shuffle(colors2)).not.toContain("chartreuse")
+    expect(shuffle(colors2)).toContain("indigo", "periwinkle", "ochre", "aquamarine", "saffron")
+  })
+})
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:34]: Good job finding the method .toContain

File: code-challenges.test.js [hunk 4]
```diff
@@ -27,31 +44,148 @@ var colors2 = ["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saf
...
+   new_colors[i] = new_colors[random_num]
+
+   // Ok, I got stuck here and it looked like in my research I would change the random number to what the current index was before it was reassigned.
+
+   new_colors[random_num] = current_color
+ }
+ return new_colors
+}
+
+// So this was returning the same array and then I realized I had the *(new_colors.length -1) outside of the Math.floor parenthesis. Once I fixed it it gave me a random output each run. Yes!
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:85]: Excellent!

File: code_challenges.rb [hunk 5]
```diff
@@ -20,15 +38,75 @@
...
+    "The #{@model} bike has #{@wheels} wheels and is going #{@current_speed} mph."
+  end
+  def pedal_faster
+    @current_speed += 10
+  end
+  def brake
+    if @current_speed >= 15
+       @current_speed -= 15
+    elsif @current_speed < 15
+          @current_speed = 0
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:102]: Okay quick question should braking always be a static number or should we let users choose how much they want to brake? What does that imply?

File: interview-questions.md [hunk 6]
```diff
@@ -7,60 +7,62 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: It is a variable that starts with an @ symbol and stores data within a class. In order to access the instance variable from outside of the class it must be stored in a method or called when using attr_accessor. One difference I found interesting between an instance variable and class variable is that instance variables can only belong to one class while class variables can be shared between a class and all of it's subclasses. As well unlike other variables, instance variables do not need to be declared. Each instance variable is actually a "dynamic append" to an object.
 
 
 
 2. What is a block in Ruby?
 
-  Your answer:
+  Your answer: A block is anonymous function that begins with a do and ends with an end. If they are written on one line they can be done using {}.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:18]: good answer!

File: interview-questions.md [hunk 7]
```diff
@@ -7,60 +7,62 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 4. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer: Object oriented programming means the code base is broken down into small sections or parts so it can be worked on without messing up other work that has been done. I am not entirely sure what the difference is between object and functional programming.
 
-  Researched answer:
+  Researched answer: Object oriented programming relies on classes and objects to structure the code so it is reusable because of the class "blueprints". Object oriented programming is also easier to debug because classes typically contain all of the applicable information. As well, because of encapsulation, the code is much more secure and limits interaction.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:36]: good work!

File: interview-questions.md [hunk 8]
```diff
@@ -7,60 +7,62 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 5. What is the difference between a class and an object?
 
-  Your answer:
+  Your answer: A class holds that data and behavior that is given to an object. It is the blueprint for an object.  
 
-  Researched answer:
+  Researched answer: A class in an abstract blueprint that is used to create more specific objects. Classes often contain broad categories while objects contain these broad pre-definitions they also allow for a more detailed or defined version of the class.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:46]: Yes! and objects made from classes are called "instances" of that class.

**Review by WalkerTexas-coder — APPROVED** [review]
Heather, as always great work! You did an amazing job with research and note taking. Your code logic makes a lot of sense. I appreciate you trying multiple approaches and learning from the differences. Keep it up!

File: code_challenges.rb (hunk 5 above)
Comment (hroulston) [reply code_challenges.rb:102]: You know it wasn't until we were going over it that I realized I mis read the instructions. I thought we needed to use the example numbers as our set inputs and outputs, then upon rereading understood that a dynamic solution was requested.

**Pull request merged**